app/src/main/java/net/sourceforge/opencamera/sensorlogging/VideoFrameInfo.java

```diff
@@ -0,0 +1,154 @@
+        );
+    }
+
+    public void submitProcessFrame(long timestamp) {
+        if (!frameProcessor.isShutdown()) {
+            frameProcessor.execute(
+                    () ->  {
+                        writeFrameTimestamp(timestamp);
+                        mFrameNumber++;
+                    }
+            );
+        } else {
+            Log.e(TAG, "Received new frame after frameProcessor executor shutdown");
+        }
+    }
+
+    public void submitProcessFrame(long timestamp, byte[] imageData, int width, int height, int rotation) {
+        if (!frameProcessor.isShutdown()) {
+            frameProcessor.execute(
+                () -> {
+                    writeFrameTimestamp(timestamp);
+                    try {
+                        if (mShouldSaveFrames && mFrameNumber % EVERY_N_FRAME == 0) {
+                            if (MyDebug.LOG) {
+                                Log.d(TAG, "Should save frame, timestamp: " + timestamp);
+                            }
+                            byte[] jpegResult = YuvImageUtils.NV21toJPEG(imageData, width, height, NV21_TO_JPEG_QUALITY);
+                            File frameFile = mStorageUtils.createOutputCaptureInfo(
+                                    StorageUtils.MEDIA_TYPE_VIDEO_FRAME, "jpg", String.valueOf(timestamp), mVideoDate
+                            );
+                            writeFrameJpeg(jpegResult, frameFile, rotation);
+                        }
+                        mFrameNumber++;
+                    } catch (IOException e) {
+                        mAppInterface.onFrameInfoRecordingFailed();
+                        Log.e(TAG, "Failed to write frame info, timestamp: " + timestamp);
+                        e.printStackTrace();
+                        this.close();
+                    }
+                }
+            );
+        } else {
+            Log.e(TAG, "Received new frame after frameProcessor executor shutdown");
+        }
+    }
+
+    private void writeFrameTimestamp(long timestamp) {
+        try {
+            mFrameBufferedWriter
+                    .append(Long.toString(timestamp))
+                    .append("\n");
+        } catch (IOException e) {
+            mAppInterface.onFrameInfoRecordingFailed();
+            Log.e(TAG, "Failed to write frame info, timestamp: " + timestamp);
+            e.printStackTrace();
+            this.close();
+        }
+    }
+
+    private void writeFrameJpeg(byte[] jpegResult, File frameFile, int rotation) throws IOException {
+        FileOutputStream fos = new FileOutputStream(frameFile);
+
+        // Apply rotation
+        Bitmap bitmap = BitmapFactory.decodeByteArray(jpegResult, 0, jpegResult.length);
+        Matrix matrix = new Matrix();
+        matrix.postRotate(rotation);
+        bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
+        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, fos);
```

Comment:
Probably, this transformation can be applied before JPEG creation to save resources from redundant compression/decompression.

Reply:
Yes, this seems irrational
Changed transformation to Bitmap (dropped usage of intermediate compressed jpeg), now it is converted directly to bitmap with Renderscript intrinsic ```yuv2rgb```, rotated, and then to jpeg which gets saved

We still need to get NV21 byte representation from Image planes but in my experiments, it was not that costly (less time than frame duration)